Auto-allocate searchable snapshots

DaveCTurner opened the pull request:
This commit allows plugins to supply their own allocator for existing shards,
generalizing the default `GatewayAllocator`. It uses this to implement a very
simple auto-allocation process for searchable snapshots which respects the
allocation deciders but which can otherwise be assigned to any node.

Relates #50999

elasticmachine commented:
Pinging @elastic/es-distributed (:Distributed/Snapshot/Restore)

DaveCTurner commented on server/src/main/java/org/elasticsearch/cluster/routing/allocation/AllocationService.java:
```diff
@@ -413,10 +421,18 @@ private void reroute(RoutingAllocation allocation) {
         // try to allocate existing shard copies first
-        gatewayAllocator.allocateUnassigned(allocation);
+        // TODO TBD the order in which allocators run depends on the order of plugins so is not properly defined. Does this matter?
+        // TODO TBD this means earlier allocators allocate shards before the later allocators can allocate possibly-higher-priority ones
```
for discussion

DaveCTurner commented on server/src/main/java/org/elasticsearch/cluster/routing/allocation/AllocationService.java:
```diff
@@ -413,10 +421,18 @@ private void reroute(RoutingAllocation allocation) {
+
         // try to allocate existing shard copies first
-        gatewayAllocator.allocateUnassigned(allocation);
+        // TODO TBD the order in which allocators run depends on the order of plugins so is not properly defined. Does this matter?
```
for discussion

ywelsch replied:
We could leave it to the core framework to decide which allocator to invoke

DaveCTurner replied:
:+1: I've moved to that model now.

DaveCTurner commented on server/src/main/java/org/elasticsearch/index/store/Store.java:
```diff
@@ -446,13 +446,8 @@ public static void tryOpenIndex(Path indexLocation, ShardId shardId, NodeEnviron
-            // Previously we called Lucene#readSegmentInfos which verifies that some Lucene metadata is readable and makes sense, but if it
-            // weren't then we would mark this shard as corrupt when allocated, so it seems that this is unnecessary (and it breaks when
-            // the shard's directory is virtual since we use SimpleFSDirectory above.
-            // TODO NORELEASE is this ok? Need to check that we definitely add a corruption marker if the metadata is corrupt.
```
With this commit, this hack is no longer needed since we do not fetch the shard info like this.

ywelsch replied:
👍

DaveCTurner commented:
@elasticmachine please run elasticsearch-ci/2

DaveCTurner commented:
I forgot to mention that of course we expect to have a more sophisticated mechanism in future with async fetching of information from the individual nodes before making a decision. This PR is only focussed on the plumbing needed to expose this kind of allocation to plugins.

DaveCTurner commented:
@elasticmachine please run elasticsearch-ci/2

henningandersen commented on x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/xpack/searchablesnapshots/SearchableSnapshotAllocator.java:
```diff
@@ -0,0 +1,117 @@
+
+        final RoutingNodes.UnassignedShards.UnassignedIterator iterator = allocation.routingNodes().unassigned().iterator();
+
+        while (iterator.hasNext()) {
```
It seems like the loop here and the handling of a NO decision belongs in the framework (allocation service?) rather than in every plugin. It is important that this is done right.

This in turn could lead to removing `GatewayAllocator` completely and letting Replica and PrimaryShardAllocator become individual `ExistingShardsAllocator`s. A method to allow `processExistingRecoveries` would be necessary though.

DaveCTurner replied:
Here we handle `NO` (and `THROTTLE`) decisions by stopping further allocation attempts, but that behaviour seems to me to be specific to this plugin (and the fact that this naive implementation checks all nodes against the deciders). For instance, the `ReplicaShardAllocator` does not share this behaviour: it only checks the deciders for nodes with at least some of the shard on disk already, and allows `NO` decisions to fall through to the `BalancedShardsAllocator` so that replicas can be allocated afresh.

I'm in two minds about iterating through the unassigned shards within each allocator. As written, we prioritise the shards that can be allocated by a plugin over the standard shards. In the case of searchable snapshots I don't think this is too bad (they should be fast to start and may well live on their own nodes). On the other hand if we pulled the loop up to the `AllocationService` then this would change the allocation order to prefer recovering higher-priority replicas instead of lower-priority primaries, which would keep the cluster in `red` health for longer after startup.

DaveCTurner replied:
Hmm having written that it's now obvious that we could solve the priority thing by sorting the shards based on whether they're a primary or not 🤦‍♂

henningandersen commented on x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/xpack/searchablesnapshots/SearchableSnapshotAllocator.java:
```diff
@@ -0,0 +1,117 @@
+
+    private static boolean isResponsibleFor(RoutingAllocation allocation, ShardRouting shardRouting) {
+        final Settings settings = allocation.metaData().getIndexSafe(shardRouting.index()).getSettings();
+        return shardRouting.unassigned() && ExistingShardsAllocator.EXISTING_SHARDS_ALLOCATOR_SETTING.get(settings).equals(ALLOCATOR_NAME);
```
If we want to name the allocator in settings, I think we should just pick the right allocator per shard from the outside?

DaveCTurner replied:
Yes, this needs the loop to be brought up to the top level, which I now see that we can do safely. I'll try that.

DaveCTurner commented on server/src/main/java/org/elasticsearch/gateway/PrimaryShardAllocator.java:
```diff
@@ -82,6 +98,14 @@ public AllocateUnassignedDecision makeAllocationDecision(final ShardRouting unas
+        final IndexMetaData indexMetaData = allocation.metaData().getIndexSafe(unassignedShard.index());
+        final String existingShardsAllocator = ExistingShardsAllocator.EXISTING_SHARDS_ALLOCATOR_SETTING.get(indexMetaData.getSettings());
+        if (unassignedShard.recoverySource().getType() == RecoverySource.Type.EXISTING_STORE
+            && existingShardsAllocator.equals(GatewayAllocator.ALLOCATOR_NAME) == false) {
```
I don't really understand this suggestion, sorry. Are you saying that we should allocate the shard as normal if the plugin is missing?

ywelsch replied:
I wonder if we should have a separate facility that ensures that certain plugins are installed in the presence of certain types of indices. Otherwise we have to build this again and again for any plugin-specific setting that we introduce.

Take for example "follower" indices (CCR). Those only make sense in the context of x-pack.

henningandersen replied:
I now see the leniency in `MetaDataIndexUpgradeService.archiveBrokenIndexSettings`, I thought we would be stricter on that. It seems dangerous that we just archive all the settings and then act like nothing happened. I think not allocating any such indices until resolved would be better, but that is likely a more intrusive change.

Maybe there is a different way we can deem indices with searchable snapshot settings without the underlying plugin more invalid than other superfluous settings?

DaveCTurner replied:
I have reverted these changes to the `PrimaryShardAllocator` and instead now the `AllocationService` chooses the correct allocator based on an index setting. I think a setting is a good approach; we could perhaps deduce the fact that a different allocator is needed for searchable snapshots via some other mechanism, but I'm not sure what that mechanism might be nor what advantage it offers.

I am generally in favour of validating that the nodes in the cluster "make sense" in terms of their plugins but that's a larger topic for another day IMO.

jasontedor replied:
@henningandersen See #26998.

henningandersen replied:
My main issue with the approach is redundancy. I can see how the redundancy here solves the issue of a missing plugin and agree that solving that is not now.

DaveCTurner commented:
I've taken this in the direction suggested by both @ywelsch and @henningandersen, added some tests, and removed the WIP label. This is ready for a proper review now.

henningandersen commented on server/src/main/java/org/elasticsearch/cluster/routing/allocation/AllocationService.java:
```diff
@@ -413,15 +427,43 @@ private void reroute(RoutingAllocation allocation) {
         assert RoutingNodes.assertShardStats(allocation.routingNodes());
     }
 
+    private void allocateExistingShards(RoutingAllocation allocation) {
```
nit: allocateExistingUnassignedShards?

DaveCTurner replied:
Ok, done in 3301a2b8444c5a37c263754033109d8bee054963.

henningandersen commented on server/src/main/java/org/elasticsearch/gateway/GatewayAllocator.java:
```diff
@@ -101,47 +106,64 @@ public int getNumberOfInFlightFetch() {
-        primaryShardAllocator.allocateUnassigned(allocation);
+    @Override
+    public void afterPrimariesBeforeReplicas(RoutingAllocation allocation) {
+        assert replicaShardAllocator != null;
```
I find this assert unnecessary since it is a final field and it is initialized in constructor?

DaveCTurner replied:
It is `null` in tests that use a subclass, and I found it all-too-easy to trip over this.

I don't really like how the tests override this class's behaviour, but don't want to start working on that kind of refactoring here.

henningandersen replied:
Ah yes, I overlooked that, thanks.

henningandersen commented on server/src/main/java/org/elasticsearch/gateway/GatewayAllocator.java:
```diff
@@ -101,47 +106,64 @@ public int getNumberOfInFlightFetch() {
+    @Override
+    public void allocateUnassigned(final RoutingAllocation allocation, ShardRouting shardRouting,
+                                   RoutingNodes.UnassignedShards.UnassignedIterator iterator) {
+        assert primaryShardAllocator != null;
```
I find these two asserts unnecessary since the fields are final and initialized in constructor?

DaveCTurner replied:
It is null in tests that use a subclass, and I found it all-too-easy to trip over this.

I don't really like how the tests override this class's behaviour, but don't want to start working on that kind of refactoring here.

henningandersen commented on x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/xpack/searchablesnapshots/SearchableSnapshotAllocator.java:
```diff
@@ -0,0 +1,109 @@
+    }
+
+    private static AllocateUnassignedDecision decideAllocation(RoutingAllocation allocation, ShardRouting shardRouting) {
+        assert shardRouting.unassigned();
```
I think the decision made here effectively bypasses the balancer. I assume you want to tackle that in a follow-up? Would be nice to know if the intention is to invoke balancer somehow here or if the code here will be changed to only consider the nodes that already have data that can be reused?

DaveCTurner replied:
Yes, the actual allocator's behaviour is not really the focus of this PR so I've just done something very simple. To be discussed in the follow-up, but indeed I agree that we will allow shards to fall through to the balancer if we can't find them a good home here.

DaveCTurner commented on server/src/main/java/org/elasticsearch/cluster/routing/allocation/AllocationService.java:
```diff
@@ -217,7 +230,8 @@ public ClusterState applyFailedShards(final ClusterState clusterState, final Lis
             }
         }
-        gatewayAllocator.applyFailedShards(allocation, failedShards);
+        existingShardsAllocators.values().forEach(existingShardsAllocator ->
```
++ done in  c9b84abb5c0baf326b626328c3acb068fd0b7e00.

DaveCTurner commented on server/src/main/java/org/elasticsearch/plugins/ClusterPlugin.java:
```diff
@@ -59,6 +60,14 @@
     }
 
+    /**
+     * Return {@link ExistingShardsAllocator} implementations added by this plugin, which will run instead of the default
```
++ done in 6cd09170c514caee8c2848fb8708ed94280555ad.

DaveCTurner commented on server/src/main/java/org/elasticsearch/cluster/routing/allocation/AllocationService.java:
```diff
@@ -70,26 +73,32 @@
+     * Inject the {@link ExistingShardsAllocator}s to use. May only be called once.
+     */
+    public void setExistingShardsAllocators(Map<String, ExistingShardsAllocator> existingShardsAllocators) {
+        assert this.existingShardsAllocators.isEmpty()
```
Ok, done in e6e5b7e617a94e725a072abe140376b8a92118ad.

DaveCTurner commented on server/src/main/java/org/elasticsearch/cluster/routing/allocation/ExistingShardsAllocator.java:
```diff
@@ -0,0 +1,87 @@
+     * Allocate any unassigned shards in the given {@link RoutingAllocation} for which this {@link ExistingShardsAllocator} is responsible.
+     */
+    void allocateUnassigned(RoutingAllocation allocation, ShardRouting shardRouting,
+                            RoutingNodes.UnassignedShards.UnassignedIterator iterator);
```
Ok, done in c7c1701c25e51874d75446fe28f8e38691f02df4. This whole area is steeped in  legacy -- there's no need to implement `Iterator` at all -- but now is  not  the time to clean that up.

DaveCTurner commented on server/src/main/java/org/elasticsearch/gateway/GatewayAllocator.java:
```diff
@@ -54,7 +55,9 @@
-public class GatewayAllocator {
+public class GatewayAllocator implements ExistingShardsAllocator {
+
+    public static final String ALLOCATOR_NAME = "gateway_allocator";
```
I prefer to keep the suffix to distinguish it from all the other things called `gateway` in this area.

DaveCTurner commented on server/src/main/java/org/elasticsearch/cluster/routing/allocation/ExistingShardsAllocator.java:
```diff
@@ -0,0 +1,87 @@
+    /**
+     * Allocate any unassigned shards in the given {@link RoutingAllocation} for which this {@link ExistingShardsAllocator} is responsible.
+     */
+    void allocateUnassigned(RoutingAllocation allocation, ShardRouting shardRouting,
```
Ok, I've moved the `ShardRouting`s first in a72d9ba05df5f3861095eb50b6450c3feae98b15.

DaveCTurner commented on server/src/main/java/org/elasticsearch/cluster/ClusterModule.java:
```diff
@@ -251,4 +252,21 @@ protected void configure() {
         bind(ShardsAllocator.class).toInstance(shardsAllocator);
     }
+
+    public void setExistingShardsAllocators(GatewayAllocator gatewayAllocator, List<ClusterPlugin> clusterPlugins) {
```
We have to defer the creation of the allocators until later, because they need a `RerouteService` and there's an (intrinsically tricky) circular dependency to resolve. We'll need to do this differently when the plugin's allocator also needs a `RerouteService` but it's premature to address that  now.

DaveCTurner replied:
Ok, addressed in 4b8888c0311d0bfc886baa7ae0349b61749d103e.

DaveCTurner commented on server/src/main/java/org/elasticsearch/cluster/routing/allocation/AllocationService.java:
```diff
@@ -470,7 +514,104 @@ protected long currentNanoTime() {
+    private ExistingShardsAllocator getAllocatorForShard(ShardRouting shardRouting, RoutingAllocation routingAllocation) {
+        assert assertInitialized();
+        final String allocatorName = ExistingShardsAllocator.EXISTING_SHARDS_ALLOCATOR_SETTING.get(
+            routingAllocation.metaData().getIndexSafe(shardRouting.index()).getSettings());
```
It involves passing the `IndexMetaData` around to a lot of places. I mean we _could_ do that if you're sure this is sufficiently expensive, but it seems surprising to me  that this would be a concern.

DaveCTurner commented:
Hey folks, any more comments on this PR?

henningandersen commented on server/src/main/java/org/elasticsearch/cluster/ClusterModule.java:
```diff
@@ -251,4 +254,21 @@ protected void configure() {
+                : clusterPlugin.getExistingShardsAllocators().entrySet()) {
+                final String allocatorName = existingShardsAllocatorEntry.getKey();
+                if (existingShardsAllocators.put(allocatorName, existingShardsAllocatorEntry.getValue()) != null) {
+                    throw new IllegalArgumentException("ExistingShardsAllocator [" + allocatorName + "] already defined");
```
Can we include `clusterPlugin.getClass().getName()` into the message?

henningandersen commented on server/src/main/java/org/elasticsearch/cluster/routing/allocation/ExistingShardsAllocator.java:
```diff
@@ -0,0 +1,124 @@
+     * Called before starting a round of allocation, allowing the allocator to invalidate some caches if appropriate.
+     */
+    void beforeAllocation(RoutingAllocation allocation);
+
```
nit: remove extra empty line.

henningandersen commented on server/src/main/java/org/elasticsearch/gateway/BaseGatewayShardAllocator.java:
```diff
@@ -45,40 +45,38 @@
-        while (unassignedIterator.hasNext()) {
-            final ShardRouting shard = unassignedIterator.next();
-            final AllocateUnassignedDecision allocateUnassignedDecision = makeAllocationDecision(shard, allocation, logger);
+    public void allocateUnassigned(RoutingAllocation allocation, ShardRouting shardRouting,
```
nit: swap argument order to make it the same as in `ExistingShardsAllocator`:
```suggestion
    public void allocateUnassigned(ShardRouting shardRouting, RoutingAllocation allocation,
```
